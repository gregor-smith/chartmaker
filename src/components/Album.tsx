import React, { FC } from 'react'
import { css, cx } from 'emotion'
import { Album as AlbumDetails, formatAlbumTitle } from '../state'


type Props = {
    details: AlbumDetails | null
    sizeRem: number
}


const baseStyle = css({
    backgroundColor: 'white',
    margin: '0.15rem'
})


export const Album: FC<Props> = ({ details, sizeRem }) => {
    const style = cx(
        baseStyle,
        css({
            width: `${sizeRem}rem`,
            height: `${sizeRem}rem`
        })
    )
    const image = details === null
        ? null
        : <img src={details.image.smallURL} alt={formatAlbumTitle(details)}/>
    return (
        <div className={style}>
            {image}
        </div>
    )
}
