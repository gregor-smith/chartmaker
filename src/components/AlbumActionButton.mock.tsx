const mock: typeof import('@/components/AlbumActionButton') = {
    AlbumActionButton: props =>
        <div {...props} className='mock-album-action-button'/>
}


export const AlbumActionButton = mock.AlbumActionButton
