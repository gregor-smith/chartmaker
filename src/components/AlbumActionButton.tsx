import { h, FunctionComponent } from 'preact'
import { css } from 'emotion'


type Props = {
    onClick: () => void
    title: string
}


const style = css({
    background: 'none',
    border: 'none'
})


export const AlbumActionButton: FunctionComponent<Props> = props =>
    <button {...props} class={style}/>
