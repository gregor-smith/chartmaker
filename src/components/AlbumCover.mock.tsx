import { cx } from 'emotion'


const mock: typeof import('@/components/AlbumCover') = {
    AlbumCover: ({
        album,
        size,
        overlayClass,
        children,
        highlighted,
        ...props
    }) => {
        const json = JSON.stringify(album)
        const className = cx('mock-album-cover', overlayClass)
        return (
            <div {...props} className={className}>
                {`Album: ${json}`}
                {`Size: ${size}`}
                {`Highlighted: ${highlighted}`}
                {children}
            </div>
        )
    }
}


export const AlbumCover = mock.AlbumCover
