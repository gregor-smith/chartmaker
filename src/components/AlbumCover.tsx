import React, { FC, DragEvent } from 'react'
import { css, cx } from 'emotion'

import { TEXT_COLOUR, ALBUM_PADDING_SIZE, highlightBackgroundStyle } from '@/style'
import { Album } from '@/types'


export type AlbumCoverProps = {
    album: Album
    size: string
    onDragStart: (event: DragEvent<HTMLDivElement>) => void
    onDragOver?: (event: DragEvent<HTMLDivElement>) => void
    onDragEnter?: (event: DragEvent<HTMLDivElement>) => void
    onDrop?: (event: DragEvent<HTMLDivElement>) => void
    onMouseEnter?: () => void
    overlayClass?: string
    highlighted?: boolean
}


const baseContainerStyle = css({
    position: 'relative',
    background: TEXT_COLOUR,
    margin: ALBUM_PADDING_SIZE
})

const imageStyle = css({
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    maxWidth: '100%',
    height: '100%',
    objectFit: 'contain'
})

// the overlay is given an additional non-emotion class so it can be used as a
// selector in the container's style
const overlaySelector = 'overlay'

const baseOverlayStyle = cx(
    overlaySelector,
    imageStyle,
    css({ zIndex: 10 })
)


export const AlbumCover: FC<AlbumCoverProps> = ({
    album,
    size,
    onDragStart,
    onDragEnter,
    onDragOver,
    onDrop,
    onMouseEnter,
    children,
    overlayClass,
    highlighted
}) => {
    const overlayStyle = cx(baseOverlayStyle, overlayClass)
    const containerStyle = cx(
        baseContainerStyle,
        css({
            width: size,
            height: size,
            [`:not(:hover) .${overlaySelector}`]: {
                display: 'none'
            }
        }),
        highlighted === false
            ? highlightBackgroundStyle
            : undefined
    )

    let image: JSX.Element | undefined
    if (!album.placeholder) {
        image = (
            // lazy loading causes problems with html2canvas so keep it default
            <img className={imageStyle}
                src={album.url}
                alt={album.name}
                crossOrigin='anonymous'/>
        )
    }

    return (
        <div className={containerStyle}
                draggable={!album.placeholder}
                onDragStart={onDragStart}
                onDragEnter={onDragEnter}
                onDragOver={onDragOver}
                onDrop={onDrop}
                onMouseEnter={onMouseEnter}
                title={album.placeholder ? undefined : album.name}>
            {image}
            <div className={overlayStyle}>
                {children}
            </div>
        </div>
    )
}
