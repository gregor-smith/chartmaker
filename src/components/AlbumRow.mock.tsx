const mock: typeof import('@/components/AlbumRow') = {
    AlbumRow: ({ children }) =>
        <div className='mock-album-row'>
            {children}
        </div>
}


export const AlbumRow = mock.AlbumRow
