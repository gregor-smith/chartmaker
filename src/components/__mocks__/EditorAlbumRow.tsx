import type { FC } from 'react'

import type { EditorAlbumRowProps } from '@/components/EditorAlbumRow'


export const EditorAlbumRow: FC<EditorAlbumRowProps> = ({ albums, size, highlighted }) => {
    const json = JSON.stringify(albums)
    return (
        <div className='mock-editor-album-title'>
            {`Albums: ${json}`}
            {`Size: ${size}`}
            {`Highlighted: ${highlighted}`}
        </div>
    )
}
