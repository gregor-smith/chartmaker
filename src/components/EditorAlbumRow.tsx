import type { FC } from 'react'

import type { Album } from '@/types'
import type { DispatchProps } from '@/reducer'
import { EditorChartAlbumCover } from '@/components/EditorChartAlbumCover'
import { getAlbumID } from '@/state'
import { AlbumRow } from '@/components/AlbumRow'


export type EditorAlbumRowProps = DispatchProps & {
    albums: Album[]
    size: string
    highlighted: number | undefined
}


export const EditorAlbumRow: FC<EditorAlbumRowProps> = ({
    dispatch,
    albums,
    size,
    highlighted
}) => {
    const covers = albums.map(album => {
        const id = getAlbumID(album)
        return (
            <EditorChartAlbumCover key={id}
                dispatch={dispatch}
                album={album}
                highlighted={highlighted === undefined
                    ? undefined
                    : id === highlighted}
                size={size}/>
        )
    })
    return (
        <AlbumRow>
            {covers}
        </AlbumRow>
    )
}
