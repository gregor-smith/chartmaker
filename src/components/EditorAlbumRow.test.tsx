import { render } from 'react-dom'

import { EditorAlbumRow } from '@/components/EditorAlbumRow'

import { RenderContainer, ignore, createTestPlaceholderAlbums } from '@/test-utils/utils'


jest.mock('@/components/EditorChartAlbumCover')
jest.mock('@/components/AlbumRow')


const container = new RenderContainer()


test('renders album row with covers', () => {
    render(
        <EditorAlbumRow dispatch={ignore}
            albums={createTestPlaceholderAlbums(5)}
            size='5rem'
            highlighted={3}/>,
        container.element
    )

    expect(container.element).toMatchSnapshot()
})
