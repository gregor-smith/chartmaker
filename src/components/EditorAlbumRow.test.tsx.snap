// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders album row with covers 1`] = `
<div
  id="test-render-container"
>
  <div
    class="mock-album-row"
  >
    <div
      class="mock-editor-chart-album-cover"
    >
      Album: 1
      Size: 5rem
      Highlighted: false
    </div>
    <div
      class="mock-editor-chart-album-cover"
    >
      Album: 2
      Size: 5rem
      Highlighted: false
    </div>
    <div
      class="mock-editor-chart-album-cover"
    >
      Album: 3
      Size: 5rem
      Highlighted: true
    </div>
    <div
      class="mock-editor-chart-album-cover"
    >
      Album: 4
      Size: 5rem
      Highlighted: false
    </div>
    <div
      class="mock-editor-chart-album-cover"
    >
      Album: 5
      Size: 5rem
      Highlighted: false
    </div>
  </div>
</div>
`;
